src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1084 @@
+                //   `Type (E1, _, ...) or `Type (F1, _, ...)`.
+                //   If there's not Type, we substitute a null check
+
+                var builder = ArrayBuilder<BoundPattern>.GetInstance();
+
+                TypeSymbol inputType = _inputType;
+                bool isEmptyPropertyPattern = node.Deconstruction.IsDefault && node.Properties is { IsDefault: false, IsEmpty: true };
+                if (_negated)
+                {
+                    if (node.DeclaredType is not null)
+                    {
+                        // `not Type`
+                        builder.Add(new BoundNegatedPattern(node.Syntax,
+                            new BoundTypePattern(node.Syntax, node.DeclaredType, node.IsExplicitNotNullTest, inputType, node.NarrowedType, node.HasErrors),
+                            inputType, narrowedType: inputType, node.HasErrors).MakeCompilerGenerated());
+                    }
+                    else if (node.InputType.CanContainNull())
+                    {
+                        // `null`
+                        BoundConstantPattern nullPattern = new BoundConstantPattern(node.Syntax,
+                            new BoundLiteral(node.Syntax, constantValueOpt: null, type: null, hasErrors: false),
+                            ConstantValue.Null, inputType, inputType, hasErrors: false);
+
+                        if (!isEmptyPropertyPattern)
+                        {
+                            nullPattern = nullPattern.MakeCompilerGenerated();
+                        }
+
+                        builder.Add(nullPattern);
+                    }
+                }
+                else
+                {
+                    if (isEmptyPropertyPattern)
+                    {
+                        builder.Add(WithInputTypeCheckIfNeeded(node));
+                    }
+                }
+
+                ImmutableArray<BoundPositionalSubpattern> deconstruction = node.Deconstruction;
+                if (!deconstruction.IsDefault)
+                {
+                    var discards = deconstruction.SelectAsArray(d => d.WithPattern(MakeDiscardPattern(d.Syntax, d.Pattern.InputType)));
+                    for (int i = 0; i < deconstruction.Length; i++)
+                    {
+                        BoundPattern rewrittenPattern = VisitWithInputTypeFromPattern(deconstruction[i].Pattern);
+                        if (CanSkip(rewrittenPattern))
+                        {
+                            continue;
+                        }
+
+                        BoundPattern rewrittenRecursive = PullBinaryPatternsOut(rewrittenPattern, inputType, makeDeconstructionPattern, args: (node, deconstruction[i], discards, i));
+                        builder.Add(rewrittenRecursive);
+
+                        if (!_negated)
+                        {
+                            inputType = rewrittenRecursive.NarrowedType;
```

Comment:
>inputType = rewrittenRecursive.NarrowedType; [](http://example.com/codeflow?start=28&length=44)

It feels like it would be better to let ```MakeBinary``` deal with adjusting the input types when it combines the patterns. It looks like ```PullBinaryPatternsOut``` does that as well.